### PR TITLE
Rama modulo3

### PR DESCRIPTION
**Pull request by oscarperezbaztan**
**model_selection.ipynb**
Para esta rama, se ha comenzado disminuyendo el tamaño del dataset del que partíamos, dado el poco espacio de RAM que paralizaba el kernel. Por ello, se ha muestreado el dataset de una forma lógica tomando pedidos random. A partir de ahí, se siguen las especificaciones del problema y se realiza un preprocesado de los datos, tomando las variables más determinantes según el EDA y comprobando los datatypes. En este sentido, se ha realizado posteriormente el train-test split, de acuerdo a la temporalidad de los datos. En un primer momento, realicé el split únicamente por grupos de usuarios, sin tener en cuenta dicha temporalidad porque había varias variables que indicaban los días desde la última compra de un producto, la variabilidad... No obstante, finalmente, dado el data leakage, he realizado la partición de acuerdo a los datos temporales por pedidos. Ordenando las fechas, el 70% de los primeros pedidos se asocian a train, el siguiente 15% a validación y el siguiente a test. Una vez realizado esto, se entrena el modelo de regresión logística con Ridge. Para ello se sigue el siguiente criterio:
- Se prueba el encoding por conteo para las variables categóricas y luego se realiza la normalización de todas las variables
- Se maximiza el rendimiento de acuerdo al área bajo la curva pr, dado que la curva ROC en este caso no ofrece información determinante dada la gran cantidad de casos TN
- Se comprueban los resultados obtenidos
ROC_Ridge
PR_Ridge
La verdad es que los resultados no son muy buenos en cuanto a la combinación recall-precisión y al comparar ambas, parece que el umbral que mejor puede funcionar en producción es el de 0.1.
th_searching_Ridge
No obstante, los resultados podrían ser mejores si se hubieran estudiado más casos, más híperparámetros...
Esta es una primera aproximación de lo que podría ser la selección del modelo. En cualquier caso, la idea es ver si los conceptos los he entendido y he realizado correctamente la separación train-test, la configuración de la pipe para normalizar de acuerdo a los datos de train o realizar el encoding basándome solo en estos datos etc... Por otro lado, he comenzado con el push_mvp pero no está del todo terminado. En cualquier caso, realizo esta pull request antes de la clase, para ver cómo he podido aplicar los conocimientos de la clase teórica a la práctica, y terminaré el script push_mvp y sus respectivos unittest.

### CODE REVIEW

File: src/module_3/evaluation_functions.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+from sklearn.model_selection import StratifiedKFold
+import matplotlib.pyplot as plt
+from datetime import timedelta
+from sklearn.model_selection import PredefinedSplit
+import pyarrow.parquet as pq
+from sklearn import metrics
+from sklearn.metrics import roc_auc_score, confusion_matrix
+
+
+def plot_precision_recall_vs_threshold(clf, X, y):
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:20]: This functions should probably not compute `model_probs`. That should be the responsability of an `inference` function if anything. This one, is about plotting precision, recall vs threshold and that's what it should do. It could take as arguments `model_prob` and `y` only.

File: src/module_3/evaluation_functions.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    plt.xlabel('Threshold Value')
+    plt.title('Precision and Recall vs Threshold')
+    plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
+    plt.show()
+
+def threshold_classification(positive_class_probs, threshold=0.5):
+    return (positive_class_probs >= threshold).astype('int')
+
+def find_best_threshold(model, X_train, y_train, X_val, y_val, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:42]: Don't Repeat Yourself. You could have a function like:
```
def predict_model_probs(model, X):
    if hasattr(model, 'predict_proba'):
        return model.predict_proba(X_train)[:, 1]
    return model.decision_function(X_train)
```
That will be reuse every time. if you need both train and val:
```
model_probs = predict_model_probs(model, X_train)
model_probs_val = predict_model_probs(model, X_val)
```

File: src/module_3/evaluation_functions.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        performance_thresholds = np.array([100 - metrics.recall_score(y_val, threshold_classification(model_probs_val, t), pos_label=1) * 100.0 for t in thresholds])
+        valid_thresholds = np.where(performance_thresholds <= 6.0)
+        best_threshold_index = valid_thresholds[0][0]
+    
+    best_threshold = thresholds[best_threshold_index]
+    print('Best threshold={:.3f}, {} on validation={:.2f}'.format(best_threshold, metric, performance_thresholds[best_threshold_index]))
+    return best_threshold
+
+def evaluate_best_threshold_test(model, threshold, X_test, y_test, metric='fscore'):
+    if hasattr(model, 'predict_proba'):
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:70]: DRY, see above

File: src/module_3/evaluation_functions.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        model_probs_val = model.decision_function(X_val)
+        
+    if (metric == 'fscore'):
+        model_precision, model_recall, thresholds = metrics.precision_recall_curve(y_train, model_probs)  
+        performance_thresholds = [metrics.f1_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
+    elif (metric == 'auc'):
+        model_fpr, model_tpr, thresholds = metrics.roc_curve(y_train, model_probs)
+        performance_thresholds = [metrics.roc_auc_score(y_val, threshold_classification(model_probs_val, t)) * 100.0 for t in thresholds]
+        best_threshold_index = np.argmax(performance_thresholds)
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:58]: Firstly, I think this is wrongly calculated. `metrics.roc_auc_score` should receive `proba` as the second argument.

Secondly, can we numerically compute what't the optimal threshold without considering the business problem? As we discussed before, the point closest to the top left corner doesn't need to be the optimal point for any specific problem. The working point is completely dependant of the problem.

File: src/module_3/evaluation_functions.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    if (metric == 'fscore'):
+        print('Fscore of the best model on test after optimizing threshold {:.2f}'.format(metrics.f1_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'recall_mean'):
+        print('Mean recall by classes of the best model on test after optimizing threshold {:.2f}'.format(metrics.roc_auc_score(y_test, threshold_classification(model_probs_test, threshold)) * 100.0))
+    elif (metric == 'fpr'):
+        print('FPR of the best model on test after optimizing threshold {:.2f}'.format(100 - metrics.recall_score(y_test, threshold_classification(model_probs_test, threshold), pos_label=1) * 100.0))
+    predictions_test_best_threshold = threshold_classification(model_probs_test, threshold)
+    print(metrics.confusion_matrix(y_test, predictions_test_best_threshold, labels=[1, 0]))
+
+def plot_validation_curve(train_scores, test_scores, param_range, ax=None, param='Param', score='Score', scale=None):
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:85]: It's not obvious at all what you are doing here. If you are plotting something that is somewhat complex you should explain what it is with docstrings.

arguments should have type hints too to make it easier to understand.

File: src/module_3/evaluation_functions.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                     color="navy", lw=lw)
+    ax.legend(loc="best")
+    
+    return ax
+
+def evaluate_classifier(clf, X, y):
+    prediction = clf.predict(X)
+    recall_mean = metrics.roc_auc_score(y, prediction) * 100
+    cm = metrics.confusion_matrix(y, prediction, labels=[1, 0])
+    recall = cm[0, 0] / cm[0, :].sum() * 100
```
Comment (guillerbf) [inline src/module_3/evaluation_functions.py:121]: This is kinda wrong. It's the recall assuming threshold is 0.5. It doesn't need to be and it shouldn't be.

File: src/module_3/push_mvp.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        
+        return data
+    except FileNotFoundError:
+        print("Error: File not found at the specified path.")
+        return None
+    except Exception as e:
+        print("An error occurred:", e)
+        return None
+
+def sample_data_by_orders(data: pd.DataFrame, sample_percentage: float) -> pd.DataFrame:
```
Comment (guillerbf) [inline src/module_3/push_mvp.py:41]: You haven't applied black. Please do

File: src/module_3/push_mvp.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    sample_orders_amount = int(len(order_ids) * sample_percentage)
+    
+    sampled_orders = np.random.choice(order_ids, size=sample_orders_amount, replace=False)
+    
+    sampled_data = data[data['order_id'].isin(sampled_orders)]
+    
+    print("Original dataset size:", data.shape)
+    print("Sampled dataset size:", sampled_data.shape)
+    
+    return sampled_data
```
Comment (guillerbf) [inline src/module_3/push_mvp.py:54]: Take a look at `pandas.sample` it would have simplified this function a lot.

File: src/module_3/push_mvp.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+def preprocess_data(data: pd.DataFrame, num_items_to_filter: int, remove_na: bool, features_to_remove: list) -> pd.DataFrame:
+
+    num_products_ordered = data.groupby('order_id')['outcome'].sum()
+    
+    orders_with_min_num_products = num_products_ordered[num_products_ordered >= num_items_to_filter].index
+    filtered_data = data[data['order_id'].isin(orders_with_min_num_products)]
+    
+    if remove_na:
+        filtered_data = filtered_data.dropna()
```
Comment (guillerbf) [inline src/module_3/push_mvp.py:64]: I'd be very careful with this type of operations. You could be dropping most of your dataset without noticing if a column starts having nans for whatever reason.

File: src/module_3/model_selection_converted.md [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+
+To address this, I devised another approach for splitting the data based on this temporal reference. This new approach involves dividing the dataset into training and testing sets while ensuring that the temporal sequence is preserved. By doing so, we aim to account for any temporal dependencies or patterns in the data, which may be critical for accurately modeling and predicting future outcomes.
+
+**NEW SPLITTING**
+
+When separating data, it's crucial to consider the date on which orders are placed for several reasons. Firstly, the temporal aspect provides valuable insights into patterns and trends over time, which can significantly impact the performance of predictive models. By incorporating temporal information into the data splitting process, we ensure that the training data represents past observations, while the testing or validation data represents future observations, mimicking real-world scenarios.
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:241]: Great!

File: src/module_3/model_selection_converted.md [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+
+Moreover, not considering the date of orders can lead to a phenomenon known as data leakage. Data leakage occurs when information from the future or unseen data is inadvertently included in the training process, leading to overly optimistic performance estimates of the model. In the context of order data, if future information, such as orders placed after a certain date, is included in the training set, the model may learn patterns that do not generalize well to unseen data, resulting in poor model performance when deployed in production.
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:243]: The key here is that you could have in `train` future behaviour that will give you extra information when predicting `test`. That information comes from the future and won't be available in production. Thus, the data leakage.

File: src/module_3/model_selection_converted.md [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+print("Testing set size:", (test_data.shape[0] / final_data_sorted.shape[0]))
+```
+
+These dates represent the end points of each subset:
+
+* The training set ends on February 9, 2021, at midnight.
+* The validation set ends on February 21, 2021, at midnight.
+* The testing set ends on March 3, 2021, at midnight.
+
+**Possible issue**: What if I have a user who makes purchases in both the training and testing sets? Will the model learn from that user? Is there a risk of data leakage?
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:295]: Nope, because that's what will happen in prod.

Comment (guillerbf) [reply src/module_3/model_selection_converted.md:295]: Anyway, a good question would be to evaluate how a model generalise to users not included in training set due to sampling, for example.

File: src/module_3/model_selection_converted.md [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+
+results = results[['param_reg__C', 'mean_test_score', 'mean_train_score']]
+
+results['param_reg__C'] = 1 / results['param_reg__C']
+
+results = results.rename(columns={'param_reg__C': 'Regularization', 'mean_test_score': 'Mean Test Score', 'mean_train_score': 'Mean Train Score'})
+results
+```
+
+The confusion matrix highlights the model's strong performance in correctly classifying non-purchases (true negatives), minimizing unnecessary notifications to users who are unlikely to engage with the product. However, the model struggles to accurately identify individuals who are likely to make a purchase (true positives), leading to missed opportunities for engagement and potential sales.
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:379]: Confusion matrix will use threshold = 0.5 which is meaningless. Cross validation can cause information leakage since it mixes past and future samples during the CV process.

File: src/module_3/model_selection_converted.md [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+The area under the precision-recall (PR) curve for the Ridge Logistic Regression model is reported as 0.16. This indicates the model's performance in terms of precision and recall, which are crucial for our sales prediction task. However, the relatively low value of 0.16 suggests that the model may struggle to effectively balance precision and recall, resulting in suboptimal performance.
+
+
+```python
+plot_PR_curve([best_estimator], X_val, y_val, ['Regresión logística Ridge'])
+```
+
+<h2 style="text-align: left;">Threshold finding</h2>
+
+The precision-recall curve for different classification thresholds indicates that achieving a balance between precision and recall would require a threshold of approximately 0.1. At this threshold, the curve suggests that we could attain a recall of approximately 0.2 and a precision of 0.2.
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:430]: A balance doesn't not need to be the optimal for a given problem. It all depends on the business case.

File: src/module_3/model_selection_converted.md [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+
+* The categorical variables, as currently encoded, do not provide significant predictive information. Further exploration is required to determine alternative encoding methods or conduct feature engineering to extract useful insights from these variables.
+
+* The ROC curve does not offer meaningful insights in this business problem. This is because the true negatives (TN) significantly outnumber the false positives (FP), resulting in a very small false positive rate (FPR), which makes it challenging to discern differences in model performance.
+
+* The precision-recall curve provides valuable information. It indicates a balanced trade-off between precision and recall at approximately 0.2 for both metrics, suggesting an optimal threshold for classification.
+
+* There is a clear need to optimize the model further. This involves exploring methods such as feature association, alternative encodings for categorical variables, and experimenting with different parameter combinations in the pipelines to uncover a more effective and optimal model.
```
Comment (guillerbf) [inline src/module_3/model_selection_converted.md:524]: Good final remarks!

**Review by guillerbf — COMMENTED** [review]
I love the description of the PR! 

Left some comments. Some are more about style and code cleanliness, while other are more fundamental. Take a look at them and incorporate the feedback for module 4
